bugsnag-plugin-android-ndk/src/main/jni/utils/serializer.c

```diff
@@ -13,100 +13,166 @@
+    ssize_t len = read(fd, event, event_size);
+    if (len != event_size) {
+        free(event);
         return NULL;
     }
-    return report;
+    return event;
 }
 
-bugsnag_report *bsg_report_v2_read(int fd) {
-    size_t report_size = sizeof(bugsnag_report);
-    bugsnag_report *report = malloc(report_size);
+bugsnag_report_v2 *bsg_report_v2_read(int fd) {
+    size_t event_size = sizeof(bugsnag_report_v2);
+    bugsnag_report_v2 *event = malloc(event_size);
 
-    ssize_t len = read(fd, report, report_size);
-    if (len != report_size) {
-      free(report);
+    ssize_t len = read(fd, event, event_size);
+    if (len != event_size) {
+      free(event);
       return NULL;
     }
-    return report;
+    return event;
 }
 
-bugsnag_report *bsg_report_read(int fd) {
+bugsnag_event *bsg_report_v3_read(int fd) {
+  size_t event_size = sizeof(bugsnag_event);
+  bugsnag_event *event = malloc(event_size);
+
+  ssize_t len = read(fd, event, event_size);
+  if (len != event_size) {
+    free(event);
+    return NULL;
+  }
+  return event;
+}
+
+bugsnag_event *bsg_event_read(int fd) {
   bsg_report_header *header = bsg_report_header_read(fd);
   if (header == NULL) {
     return NULL;
   }
 
-  int report_version = header->version;
+  int event_version = header->version;
   free(header);
+  bugsnag_event *event = NULL;
+
+  if (event_version == 1) { // 'event->unhandled_events' was added in v2
+    bugsnag_report_v1 *report_v1 = bsg_report_v1_read(fd);
+    event = map_v1_to_report(report_v1);
+  } else if (event_version == 2) {
+    bugsnag_report_v2 *report_v2 = bsg_report_v2_read(fd);
+    event = map_v2_to_report(report_v2);
+  } else {
+    event = bsg_report_v3_read(fd);
+  }
+  return event;
+}
 
-  bugsnag_report *report = NULL;
-
-  if (report_version == 1) { // 'report->unhandled_events' was added in v2
-      bugsnag_report_v1 *report_v1 = bsg_report_v1_read(fd);
-
-      if (report_v1 != NULL) {
-          report = malloc(sizeof(bugsnag_report));
-
-          report->notifier = report_v1->notifier;
-          report->app = report_v1->app;
-          report->device = report_v1->device;
-          report->user = report_v1->user;
-          report->exception = report_v1->exception;
-          report->metadata = report_v1->metadata;
-          report->crumb_count = report_v1->crumb_count;
-          report->crumb_first_index = report_v1->crumb_first_index;
-
-          size_t breadcrumb_size = sizeof(bugsnag_breadcrumb) * BUGSNAG_CRUMBS_MAX;
-          memcpy(&report->breadcrumbs, report_v1->breadcrumbs, breadcrumb_size);
-
-          strcpy(report->context, report_v1->context);
-          report->severity = report_v1->severity;
-          strcpy(report->session_id, report_v1->session_id);
-          strcpy(report->session_start, report_v1->session_start);
-          report->handled_events = report_v1->handled_events;
-          report->unhandled_events = 1;
+bugsnag_event *map_v2_to_report(bugsnag_report_v2 *report_v2) {
```

Comment:
All of these functions should use the `bsg_` prefix, to avoid collisions with any other libraries running in the host app.

Reply:
Prefixed these methods.